confgenerator/feature.go

```diff
@@ -0,0 +1,246 @@
+package confgenerator
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+// Errors returned by ExtractFeatures can be tested against these errors using
+// errors.Is
+var (
+	ErrTrackingInlineStruct   = errors.New("cannot have tracking on inline struct")
+	ErrTrackingOverrideStruct = errors.New("struct that has tracking tag must not be empty")
+	ErrMapAsField             = errors.New("map type for a field is not supported")
+	ErrTrackingOverrideMap    = errors.New("map that has tracking tag must not be empty")
+	ErrInvalidType            = errors.New("object in path must be of type Component")
+)
+
+type Feature struct {
+	// Module defines the sub-agent: metrics or logging
+	Module string
+	// Kind defines the kind: receivers or processors
+	Kind string
+	// Type from Component.Type()
+	Type string
+	// Key: set of keys that will be joined together for feature tracking metrics
+	Key []string
+	// Value defined from fields of UnifiedConfig.
+	Value string
+}
+
+// ExtractFeatures fields that containing a tracking tag will be tracked.
+// Automatic collection of bool or int fields. Any value that exists on tracking
+// tag will be used instead of value from UnifiedConfig.
+func ExtractFeatures(uc *UnifiedConfig) ([]Feature, error) {
+	allFeatures := getOverriddenDefaultPipelines(uc)
+	f, err := trackingFeatures(reflect.ValueOf(uc), metadata{}, Feature{})
+	if err != nil {
+		return nil, err
+	}
+	allFeatures = append(allFeatures, f...)
+	return allFeatures, nil
+}
+
+func trackingFeatures(c reflect.Value, m metadata, feature Feature) ([]Feature, error) {
+	if m.IsExcluded {
+		return nil, nil
+	}
+	t := c.Type()
+
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+
+	v := reflect.Indirect(c)
+	if v.Kind() == reflect.Invalid || v.IsZero() {
+		return nil, nil
+	}
+
+	var features []Feature
+
+	switch kind := t.Kind(); {
+	case kind == reflect.Struct:
+		// If struct is inline there is no associated name for key generation
+		// By default inline structs of a tracked field are also tracked
+		if m.IsInline && m.HasTracking {
+			return nil, ErrTrackingInlineStruct
+		}
+
+		// If struct has tracking it must have a value
+		if m.HasTracking && !m.HasOverride {
+			return nil, ErrTrackingOverrideStruct
+		}
+
+		// Path for execution is coupled with the UnifiedConfig structure, feature
+		// gets populated as we iterate through each level of the struct
+		if feature.Module == "" {
+			feature.Module = m.FieldTag
+		} else if feature.Kind == "" {
+			feature.Kind = m.FieldTag
+		} else if !m.IsInline {
+			feature.Key = append(feature.Key, m.FieldTag)
+		}
+
+		// For structs that in a Component
```

Comment:
Do you mean `For structs that are in a Component`?

See my other comment about assuming depth of recursion. Are we using `feature.Module != "" && feature.Kind != "" && feature.Type != "" && feature.Key != nil` to determine whether we are in a struct that is defined inside a receiver/processor component? 
Also what `Type` will these structs have? None?

Reply:
Changed in: [237e331](https://github.com/GoogleCloudPlatform/ops-agent/pull/853/commits/237e331081f573c60c5da6d731121d7f4cd56a46)